sycl: Add more debug prints

Rbiessy opened the pull request:
This adds more debugging prints related to the operations being run. `scope_op_debug_print` is introduced to print the "call ..." and "call ... done" logs in a single line.
This adds more information for all the operations and its destination and input tensors: the tensor type, its number of elements, stride sizes and whether the tensor is strided (i.e. non-contiguous) or permuted.
For mul_mat this also add some debug prints when conversations kernels are called from and to fp16, fp32 and quantized types.
This also adds debugging prints for a few operations that were missing it.

The output for llama-2-7b.Q4_0 PP and TG is added here as an example:
* [llama-2-7b.Q4_0__log_bench_pp_512_no_warmup.txt](https://github.com/user-attachments/files/20389528/llama-2-7b.Q4_0__log_bench_pp_512_no_warmup.txt)
* [llama-2-7b.Q4_0__log_bench_tg_1_no_warmup.txt](https://github.com/user-attachments/files/20389538/llama-2-7b.Q4_0__log_bench_tg_1_no_warmup.txt)

Rbiessy committed:
sycl: Add more debug prints

NeoZhangJianyu commented on ggml/src/ggml-sycl/cpy.cpp:
```diff
@@ -694,8 +694,6 @@ void ggml_sycl_cpy(ggml_backend_sycl_context & ctx, const ggml_tensor * src0, co
 void ggml_sycl_dup(ggml_backend_sycl_context & ctx, ggml_tensor * dst) {
-    // TODO: why do we pass dst as src1 here?
-    GGML_SYCL_DEBUG("[SYCL] call %s\n", __func__);
+    scope_op_debug_print scope_dbg_print(__func__, dst, /*num_src=*/1);
```
I have a concern that this dynamic method will decrease the performance of running time:

The legacy method won't impact the performance when disable the debug mode.
But this method will spend time to create the object when disable the debug mode.

Have you test the performance impact with legacy code when disable the debug mode?

NeoZhangJianyu replied:
`scope_op_debug_print scope_dbg_print(__func__, dst, /*num_src=*/1);` 

this code will trigger constructor function when entry ggml_sycl_dup().
When call ggml_sycl_dup() n times, the constructor function will be called n times.

Legacy code is the static code. It won't take more time when entry ggml_sycl_dup().

Rbiessy replied:
AFAIK the constructors defined in the headers can be inlined by the compiler if it chooses to do so in which case it will just end up with the same `if (!g_ggml_sycl_debug)` check. But yeah I will verify this with a few becnhmarks.

Rbiessy replied:
Since I added quite a few debugging prints I thought it could be a good opportunity to introduce `__builtin_expect` in `LIKELY` and `UNLIKELY` macro as it's often used. This can let the optimizer choose the most likely path in branches and better pipeline the instructions.
I compared the performance with a small model on PVC where the host should have be the biggest bottleneck. It looks good to me:

Master:
| model                          |       size |     params | backend    | ngl | threads |    sm | mmap |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ----: | ---: | --------------: | -------------------: |
| qwen2 1.5B Q4_0                | 1013.62 MiB |     1.78 B | SYCL       |  99 |       8 |  none |    0 |           pp512 |      5172.90 ± 24.07 |
| qwen2 1.5B Q4_0                | 1013.62 MiB |     1.78 B | SYCL       |  99 |       8 |  none |    0 |           tg128 |        159.39 ± 0.52 |

PR:
| model                          |       size |     params | backend    | ngl | threads |    sm | mmap |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ----: | ---: | --------------: | -------------------: |
| qwen2 1.5B Q4_0                | 1013.62 MiB |     1.78 B | SYCL       |  99 |       8 |  none |    0 |           pp512 |      5179.98 ± 21.60 |
| qwen2 1.5B Q4_0                | 1013.62 MiB |     1.78 B | SYCL       |  99 |       8 |  none |    0 |           tg128 |        159.55 ± 0.42 |

NeoZhangJianyu replied:
Yes, this test result shows there is less impact to performance.

Entry the OP, construct a class instance; before exit, deconstruct the instance.
The C++ code will take more time than static code in running time.

The absolutely performance gap maybe very small, like 0.000001ms.
Even if the running time is increased 1000 times, the absolutely increased time is still small.

But from code design viewpoint, I don't like this solution.
We could use more efficient code, instead of good C++ style but with lower efficiency.

It's only suggestion. It won't block to merge this PR if you don't accept.

If performance is lost by many pieces of small code, it's very hard to get it back.

qnixsynapse reviewed (commented):
Nice.
Is it possible to extend this to backend interface functions as well such as init, set, memset, clear, reset, etc?

Rbiessy committed:
Add more prints

Rbiessy committed:
Use __builtin_expect

Rbiessy committed:
clang-format

Rbiessy commented:
> Nice. Is it possible to extend this to backend interface functions as well such as init, set, memset, clear, reset, etc?

I added some more logs and updated the example in the PR description. Note I am not reusing `scope_op_debug_print` everywhere as I wanted to make the difference between logs coming from the llama backend (like the functions you mentioned) and logs coming from the model (like operations that are run). I added some `[OP]` tag in the log to distinct between the 2.

Alcpz reviewed (approved):
LGTM. If logs get too noisy we can implement a couple of levels to the debug variable in the future to either print ops or API calls.

qnixsynapse reviewed (approved):
LGTM

Rbiessy committed:
Use string_view

Rbiessy commented:
I switched from using `std::string` to `std::string_view` to avoid any cost associated with `std::string`. As I measured the PR wasn't enough to impact performance but a small reproducer showed more instructions generated. With `std::string_view` that is not the case anymore, see `f1` (using macro) and `f2` (using `scope_dbg_print`) are identical: https://gcc.godbolt.org/z/5T84n5rzs
I'll wait until Monday to merge this if there are no other comments.
